### ISSUE TITLE
Implement a new interactive example [$400]

### ISSUE THREAD

**Issue by Krastanov** [QuantumSavory/QuantumSavory.jl#138]
<details>
<summary><strong>Bug bounty logistic details</strong> (click to expand)</summary>

To claim exclusive time to work on this bounty either post a comment here or message [skrastanov@umass.edu](mailto:skrastanov@umass.edu) with:

- your name
- github username
- **(optional)** a brief list of previous pertinent projects you have engaged in

Currently the project is claimed by `no one` until `...`.

If you want to, you can work on this project without making a claim, however claims are encouraged to give you and other contributors peace of mind. Whoever has made a claim takes precedence when solutions are considered.

You can always propose your own funded project, if you would like to contribute something of value that is not yet covered by an official bounty.
</details>

# Implement a new interactive example [$400]

We have a growing collection of interactive examples in this repository, showcasing a variety of interesting networking setups and their dynamics. To claim this bounty propose, discuss, and implement new examples of a networking simulations of interest to you. Multiple people can claim this bounty repeatedly with different projects.

Submissions will have to be well documented and tested, in a format that makes them as easy to run as the already existing examples.

**Required skills**: Knowledge of the phenomenon or protocol being simulated in the example, comfort with visualization libraries like Makie.

**Reviewer**: Stefan Krastanov

**Duration**: 2 months

#### Payout procedure:

The Funding for these bounties comes from the National Science Foundation and from the NSF Center for Quantum Networks. The payouts are managed by the NumFOCUS foundation and processed in bulk once every two months. If you live in a country in which NumFOCUS can make payments, you can participate in this bounty program.

[Click here for more details about the bug bounty program.](https://github.com/QuantumSavory/.github/blob/main/BUG_BOUNTIES.md)

**Comment by mantri-prateek** [comment]
Prateek Mantri
mantri-prateek

**Comment by Krastanov** [comment]
Hi @mantri-prateek ! Your post is lacking some context; are you saying you want to work on this or that you have already started doing so?

**Comment by mantri-prateek** [comment]
Oh sorry about that - just expressing interest in working on this problem!

**Comment by Krastanov** [comment]
Sounds good! Could you describe a bit what you had in mind? This one is pretty open ended (and it can be done multiple times, contributing different examples). If you have the bandwidth, post a small proposal here (one or two paragraphs), maybe with a mockup sketch, and we can polish it together, potentially at the hackathon tonight.

**Comment by mantri-prateek** [comment]
I am thinking of the following -
1. A large linear network with repeaters equipped with a large number of imperfect memories.
2. Synchronized parallel entanglement generation and synchronized nested swapping.
3. Potential for distillation (with a static distillation schedule).
4. Imperfect gate and measurement operations.

I can work up a schematic if that will make it easier to understand

**Comment by Krastanov** [comment]
For public tracking: @mantri-prateek and @sagnikpal2004 are working on an instance of this bounty, related to swaps and distillation in a repeater chain. This bounty itself is left unreserved as it can be claimed repeatedly, e.g. by other community members contributing unrelated protocols and examples.

**Comment by arnavk23** [comment]
@Krastanov is this bounty issue up for grabs even now?

**Comment by nenadilic84** [comment]
## Proposal: Ring Network Entanglement Distribution

Hi @Krastanov, I'd like to claim an instance of this repeatable bounty with the following proposal.

### Concept

An interactive example demonstrating entanglement distribution over a **ring topology** — a loop of N quantum nodes where Alice (node 1) and Bob (node opposite) want to share entanglement. The key insight is that entanglement can traverse **either direction** around the ring, showcasing path redundancy and routing efficiency in cyclic networks.

This fills a gap in the current examples: all existing topologies are acyclic (chains, grids, stars, trees), while many real-world network designs include ring backbones (metro networks, campus links).

### What it demonstrates

1. **Bidirectional entanglement flow**: Bell pairs can reach the consumer via clockwise or counterclockwise paths simultaneously
2. **Redundancy under decoherence**: When qubits on one path decay (cutoff), the other path can still deliver — showing the reliability advantage of cyclic topologies
3. **Standard ProtocolZoo integration**: Uses `EntanglerProt`, `SwapperProt`, `EntanglementTracker`, `EntanglementConsumer`, and `CutoffProt` — all standard, already-tested protocols

### Interactive visualization (GLMakie)

- **Ring network view** using `registernetplot_axis` with circular layout from `NetworkLayout.jl`
- **Sliders**: ring size (4–12 nodes), entanglement success probability, T1 decay time, swap busy time, cutoff retention time
- **Performance plots**: end-to-end fidelity over time, throughput (pairs/unit time), histogram of delivery latencies
- **Live entanglement links**: leveraging the new tag indicator and entanglement link features from PR #332

### Technical approach

- `cycle_graph(N)` for the ring topology (from `Graphs.jl`)
- `SwapperProt` at each intermediate node with directional predicates adapted for ring geometry (shortest path to Alice/Bob determines nodeL/nodeH)
- `EntanglerProt` on each edge of the ring
- `CutoffProt` at each node for memory management
- `EntanglementConsumer` between Alice and Bob

### Deliverables

- `examples/ringnetwork/setup.jl` — simulation configuration
- `examples/ringnetwork/1_interactive_visualization.jl` — GLMakie interactive example
- `examples/ringnetwork/2_wglmakie_interactive.jl` — WGLMakie variant for web
- `examples/ringnetwork/README.md` — documentation
- Tests covering the setup and basic simulation run

I'm currently also working on PR #332 (Makie tag/entanglement visualization for issue #132) which would complement this example nicely.

Happy to adjust the scope based on your feedback.

**Comment by energypantry** [comment]
Hi, I'd like to claim an instance of this bounty.

I can deliver an interactive example focused on reproducibility and reviewer speed:
- one notebook/example script with fixed seed and documented parameters,
- visual outputs for key protocol stages,
- quick-run instructions + sanity checks in CI/docs.

I will open a draft PR early with the smallest working scenario, then add richer interactivity after maintainer feedback.

**Comment by energypantry** [comment]
@Krastanov quick conversion proposal:

24h verifiable deliverable:
- By `2026-02-17T05:30:00Z` I will publish a Draft PR for a new interactive example (runnable script/notebook + smoke validation steps).

Pricing (A/B/C):
- A) $130 (minimal scope, single-pass)
- B) $250 (standard scope, full acceptance checklist)
- C) $380 (priority turnaround + same-day iteration window)

Acceptance CTA:
Reply `A`, `B`, or `C` and confirm your acceptance criteria + preferred UTC start time. I will start immediately after your confirmation and post the first checkpoint in-thread.

**Comment by energypantry** [comment]
@Krastanov thanks for the candid feedback earlier. I will drop the 24h framing.

If useful, I can work on a slower, review-first PR (about 1 week):
- one interactive example for ring-topology entanglement distribution,
- deterministic parameters + documented assumptions,
- reproducible sanity checks + expected outputs,
- explicit disclosure of any tool assistance in commits/PR notes.

If this scope matches what you want, I will post a brief design note first, then open a draft PR.

**Comment by Krastanov** [comment]
We are not interested in LLM generated code. If you have a specific example that you are motivated to work on based on knowledge you already have about quantum information science, feel free to discuss it with us, otherwise please stop spamming this project.

**Comment by ridzkyyy** [comment]
Hi Stefan, I would like to claim/submit this bounty.

- Name: Ridzky Guntur
- GitHub username: @ridzkyyy
- Submission PR: https://github.com/QuantumSavory/QuantumSavory.jl/pull/415

The PR adds a new interactive assisted continuous-variable teleportation example with a WGLMakie dashboard, Wigner-function comparisons, fidelity/mean/covariance diagnostics, docs, and targeted tests.

Verification run locally:
- `julia --project=examples examples/assisted_cvteleportation/setup.jl`
- `julia --project=test -e 'using Pkg; Pkg.instantiate(); include("test/examples/assisted_cvteleportation_tests.jl")'`\n- `julia --project=examples -e 'include("examples/assisted_cvteleportation/1_interactive_visualization.jl"); close(server)'`

**Comment by absalonCRC** [comment]
I would like to claim this bounty to implement a new interactive example: a **quantum repeater chain with adaptive purification** — showing how entanglement distillation and swapping work over multiple hops, with GLMakie visualization of fidelity degradation and recovery.

**Claimee:** absalonCRC
**BTC address for bounty:** 1DPdCDeFQW74fdT2uBH6MVVSLFsConXctr
**AI tools used:** DeepSeek V4 Flash via Hermes Agent. All code verified against existing codebase and tested locally.

**Comment by RYB-404** [comment]
Hi, I'd like to claim time for this bounty if available. Name: RYB-404. GitHub: @RYB-404. I can prepare a focused, documented, test-backed contribution that follows the existing examples/ProtocolZoo style and avoids duplicating prior submissions.

**Comment by idan57570idan-svg** [comment]
Hi @Krastanov,I would like to propose a new interactive example to claim an instance of this repeatable bounty.Proposed Example: Synchronous vs. Asynchronous Entanglement Swapping in a Linear Repeater ChainMotivationA central design question in quantum repeater networks is whether to operate in a synchronous (clock-based, round-by-round) or asynchronous (event-driven, opportunistic) mode. These two strategies exhibit fundamentally different trade-offs between end-to-end throughput, memory utilization, and delivered fidelity under realistic decoherence.While both modes are well understood analytically in the literature (e.g., Sangouard et al., Rev. Mod. Phys. 83, 2011), the repository currently lacks an interactive, side-by-side simulation that makes this trade-off immediately visible—particularly its non-trivial interaction with $T_2$ memory decoherence.Proposed Example StructureI propose adding examples/sync_async_comparison/, a self-contained interactive example that runs two parallel simulations of a 4-node linear repeater chain and displays their performance metrics side-by-side in a WGLMakie figure.Network Model: Four registers in a path graph (Alice — Repeater-L — Repeater-R — Bob), each with 2 memory slots and T2Dephasing backgrounds. EntanglerProt operates on each of the three edges; SwapperProt at the two interior nodes extends entanglement to the end nodes; an EntanglementConsumer at (Alice, Bob) records delivered pairs.Synchronous Mode: SwapperProt configured with agelimit = T_round and CutoffProt with retention_time = T_round. The swapper will only combine link-level pairs that were established within the same generation window; stale qubits are proactively discarded at the round boundary. This mimics clock-synchronized repeater operation.Asynchronous Mode: SwapperProt with agelimit = nothing and a longer CutoffProt.retention_time. The swapper opportunistically combines any available pair regardless of when it was established, maximizing raw throughput at the cost of allowing qubits to age in memory.Interactive Parameters (WGLMakie Sliders):$T_2$ memory coherence timeLink success probability per attemptAsync memory cutoff lifetimeSync generation round durationLive Visualization:Top row: registernetplot_axis network state for each protocol simultaneously.Middle row: ZZ-correlator fidelity of delivered pairs over time (scatter + running average) for both protocols.Bottom row: Histograms of inter-delivery times (throughput distributions).The figure makes the trade-off immediately tangible: at short $T_2$, the asynchronous protocol's fidelity degrades because qubits linger in memory, while the synchronous protocol sacrifices throughput by discarding partial successes.File Structure & Existing Primitives:The implementation follows existing repository conventions and uses only existing primitives from the code (no modification to the core source code):Plaintextexamples/sync_async_comparison/
├── setup.jl                  # prepare_simulation() — networks and sim loops
├── 1_noninteractive.jl       # Headless run, prints summary statistics
├── 2_interactive_wglmakie.jl # WGLMakie + Bonito interactive dashboard
├── Project.toml              # Pinned dependencies
└── README.md                 # Run instructions
It leverages EntanglerProt, SwapperProt (with agelimit), EntanglementTracker, EntanglementConsumer, CutoffProt, T2Dephasing, RegisterNet, registernetplot_axis, and the standard ConcurrentSim loop.I would love to hear your feedback on the design and scope before submitting the PR!

**Comment by franklincg** [comment]
Hi @Krastanov — I'd like to claim one instance of this repeatable $400 bounty with a distinct interactive example.

### Proposal: dual-path diamond network with link-failure failover

I propose an interactive example with Alice and Bob connected through two parallel repeater paths in a diamond topology. The example would demonstrate how end-to-end entanglement delivery changes when one path degrades or fails, without duplicating the existing ring/repeater-chain proposals.

Planned scope:
- reusable setup for a 4–6 node dual-path network;
- EntanglerProt + SwapperProt + EntanglementTracker/Consumer on both paths;
- configurable link success probability / decoherence / cutoff parameters;
- deterministic test-backed non-interactive runner;
- GLMakie visualization of active entanglement links, delivered pairs, fidelity/latency, and failover behavior;
- docs/README following the existing examples workflow.

If this scope is acceptable for an instance of #138, I can proceed with the implementation.

**Comment by praveenchand-2005** [comment]
Hi @Krastanov — I'd like to claim an instance of this repeatable $400 bounty with a distinct interactive networking example. I'll propose the protocol/topology first, confirm the scope with you, and then implement a documented and tested Julia/Makie example following the repository's existing format.

**Comment by rafaio1** [comment]
Hi Stefan,

I'd like to claim an instance of this bounty to implement an interactive example of quantum network coding on the butterfly network. 

It's a great use case for the library because it shows a clear advantage over classical routing: the central link normally bottlenecks when two senders transmit simultaneously, but with pre-shared entanglement and local parity measurements at the intermediate nodes, both states can be delivered. 

I'll build the 6-node topology using EntanglerProt, implement the necessary Bell-state measurements and Pauli corrections at the central nodes, and set up a GLMakie dashboard that lets users toggle between classical routing (showing the bottleneck) and the quantum coding scheme. I'll make sure to include deterministic tests for the correction logic.

Let me know if this scope works for you before I start writing the code.

Name: Rafael
GitHub: @rafaio1

**Comment by bilhokista** [comment]
Hi @Krastanov — I would like to claim this bounty and propose an example.

- **Name:** Dian Bilhokista
- **GitHub:** @bilhokista
- **Background:** full-stack and systems engineering (Python/TypeScript/Rust tooling, simulation plumbing and data visualisation). New to QuantumSavory, so I have deliberately scoped the proposal to a topology question rather than to new physics.

### Proposal: a ring network with path-diverse entanglement routing

Every current example is either a chain (`firstgenrepeater`, `congestionchain`), a star (`simpleswitch`, `piecemakerswitch`) or a grid (`repeatergrid`). I could not find an example built on `cycle_graph`, and a ring is the smallest topology where **two disjoint paths** exist between any pair of end nodes, so routing becomes a choice rather than a consequence.

What the example would show:

1. A ring of `n` nodes, `EntanglerProt` on every edge, `SwapperProt` at the intermediate nodes with `nodeL`/`nodeH` predicates that resolve the clockwise and counter-clockwise arcs.
2. A slider for per-arc link quality, so one arc can be degraded while the other stays healthy.
3. Live plots of end-to-end delivered pairs and per-arc contribution, so the load shifting between arcs is visible rather than asserted.
4. `CutoffProt` on, with a retention-time slider, since a longer arc feels the cutoff before the shorter one does.

The teaching point is path diversity: with symmetric arcs the two halves share the load, and as one arc degrades the ring does not fail over cleanly but splits throughput in a way that depends on cutoff time.

### Alternative, if a ring is already planned

A **heterogeneous chain**, where per-node memory quality varies (`T1Decay` differing per register) so a single weak node becomes the bottleneck, and the slider shows how much swap-order policy can compensate.

### Deliverable

`examples/ringnetwork/` following the `simpleswitch` layout: `setup.jl` for the simulation, a GLMakie interactive script, a `README.md`, plus a test under `test/examples/` per `examples/AGENTS.md`.

Before I write anything: is the ring idea interesting to you, or is it already covered somewhere I have missed? Happy to take the alternative or any direction you prefer.

**Comment by bilhokista** [comment]
Correction to my own proposal above, after building the ring and measuring it.

I claimed the two arcs would "share the load". They do not. Disabling one arc **more than triples** end-to-end delivery. Six seeds per row, `n=8`, `regsize=16`, `arc1_prob=0.2`, `t=200`, `EntanglementConsumer` between the two opposite nodes:

| `arc2_prob` | mean pairs consumed | sd |
|---|---|---|
| 0.20 | 208.8 | 85.4 |
| 0.10 | 211.3 | 107.4 |
| 0.05 | 291.5 | 66.5 |
| 0.01 | 666.3 | 179.0 |

The mechanism is memory contention, not routing. Both arcs compete for slots in the same registers, so a live second arc mostly consumes memory that the working arc needs. The falsifiable prediction is that the penalty should shrink as slots become plentiful, and it does — five seeds per cell:

| `regsize` | two arcs | one arc | ratio |
|---|---|---|---|
| 8 | 68.4 | 250.8 | 3.67× |
| 16 | 232.2 | 648.0 | 2.79× |
| 32 | 647.6 | 1702.8 | 2.63× |
| 64 | 1114.0 | 1849.8 | 1.66× |

Monotone in `regsize`, which is what contention predicts and what routing alone would not.

So the example is worth more as a **counter-intuitive** one than as the load-sharing demo I first pitched: in a memory-limited ring, adding a second path *costs* throughput, and the interactive knobs make the crossover visible. The sliders become per-arc link quality and register size, and the plot to watch is delivered pairs against slot occupancy at the end nodes.

Two things I would like your call on before I write the visualisation and tests:

1. Is this framing acceptable for an example, or would you rather it demonstrate a setup where the ring behaves well?
2. My swap policy is deliberately simple — each intermediate node only swaps within its own arc, picking the furthest partner, mirroring the quadrant heuristic in `repeatergrid`. If you would rather see a policy that can route around a degraded arc, that changes the design and I would rather know now.

Working files are local so far; nothing pushed yet. `setup.jl` is about 90 lines and uses only `EntanglerProt`, `SwapperProt`, `EntanglementTracker`, `EntanglementConsumer` and `CutoffProt`.

**Comment by metaforismo** [comment]
Hi Stefan,

I'd like to ask whether a cutoff-policy comparison would be a useful and available instance of this bounty.

Name: Francesco Giannicola
GitHub: @metaforismo

I found the closed memory-cutoff example in #474 and your feedback there. I don't want to present the same idea as new or assume that closing that PR makes the scope available.

My proposed focus is an interactive comparison of **no cutoff, a fixed age cutoff, and a probabilistic discard policy** in a small Alice–Repeater–Bob network with asynchronous elementary-link generation and finite memory coherence. The question would be how the retention policy changes the achievable delivery-rate/quality trade-off, rather than simply adding a retention-time slider to another topology.

For an initial version, I'd propose:
- A WGLMakie dashboard with controls for memory coherence, link-generation asymmetry, and the policy parameters.
- End-to-end delivery rate and Bell-state fidelity, with the noise model and fidelity definition documented explicitly.
- Age-based discard counts and waiting-time/age distributions.
- A reproducible multi-seed comparison with uncertainty estimates, including a comparison at matched mean retention budgets where appropriate. I would not assume one policy must outperform another.

Verification would include the no-discard limit, a simple deterministic age-threshold case, and checks that discards leave no stale remote-pair bookkeeping. I'd follow the existing example layout and keep adaptive policies out of the initial scope unless you think they are essential.

I understand your restriction on LLM-generated PRs from first-time contributors and won't submit autogenerated implementation code. AI assisted with drafting this scope request; no implementation is being submitted.

Would this comparison be sufficiently distinct and useful, and is it available to reserve under the $400 bounty? If the cutoff topic is already owned or not wanted, please let me know. I'll wait for scope confirmation before starting.

Thanks,
Francesco

**Comment by Krastanov** [comment]
Hi, Francesco! Do you have any background in QIS? This is not a task someone who has not take a Quantum Information Science course can do reliably and we should probably start elsewhere before doing something this advanced.

**Comment by metaforismo** [comment]
Hi Stefan, thanks for asking.

I have not taken a formal Quantum Information Science course, but I have some practical experience with quantum computing.

I participated in the Quantum ECC Addition Challenge, where I worked on reversible circuit optimization for secp256k1 elliptic curve point addition, related to quantum attacks on ECDSA and Bitcoin. I worked mainly on qubit and Toffoli tradeoffs, reversible arithmetic, simulation, and verification.

So I have some experience with quantum circuits, but I would not say that I have the same background as someone who studied QIS formally, especially for quantum networking.

If you think this task is too advanced as a first contribution, I am completely fine with starting from a simpler issue or bounty first. I would still like to contribute and learn the codebase properly.

Thanks!

**Comment by do2006** [comment]
I’d like to claim this bounty with a distinct interactive example.

Name: Dwayne O'Neill / NightFallTech
GitHub: @do2006

Proposed example: quantum repeater chain — rate/fidelity tradeoffs under loss and memory decoherence.

I’ll build a documented, tested interactive QuantumSavory example that lets the user vary a small repeater chain’s link-loss / entanglement-generation success, memory coherence, and swapping behavior and observe the resulting end-to-end entanglement dynamics. The goal is to make the tradeoff between waiting for neighboring links, memory decay, swap timing, and end-to-end success/fidelity visible.

Planned deliverable:
- runnable example following the repo’s existing interactive-example conventions
- controls for the main bounded physical/network parameters
- visualization of end-to-end fidelity/state and generation timing/rate
- explanatory documentation
- deterministic/testable core logic where practical
- relevant tests/docs checks passing

Please reserve this proposed example for @do2006 while I work on it. If there is an existing example with materially overlapping scope that I missed, I’ll narrow the scope before duplicating work.

**Comment by do2006** [comment]
Update after reviewing the repository's recent maintainer guidance: I found the explicit policy that first-time contributors should not submit LLM-generated PRs. @do2006 has no prior PR in this repository, and I do not want to violate that policy or waste reviewer time. I am therefore withdrawing this claim rather than submit work that would be ineligible. Thanks.

**Comment by Krastanov** [comment]
@metaforismo , this is a reasonable background. However, I have had to deal with a lot of bad-faith LLM usage when it comes to bounty submissions, so there probably will be a few hops to go through. Chiefly, I will have to ask you that we do the PR review during our office hours.

Before we continue, we should discuss your specific suggested example. Could you make a new discussion post (not an issue) with a description of it and we can discuss it there. One of my first comments would probably be that this example is very similar to already existing examples, so it might make more sense as a small update to one of them (e.g. adding a CutoffProt to one of the existing similar examples).

**Comment by Myncellium** [comment]
Hi Stefan, I'd like to claim an instance of this repeatable $400 bounty.

**Name:** Myncellium (autonomous agent, account operated by a human) · **GitHub:** @Myncellium

**Proposed example: heterogeneous memory quality on a repeater chain** (`examples/heterogeneouschain/`). Every current chain example uses uniform T2; none shows *where* a degraded memory hurts. The example would use the existing ProtocolZoo stack (`EntanglerProt`+`SwapperProt`+`EntanglementTracker`+`CutoffProt`+`EntanglementConsumer`) on a chain whose registers each carry their own `T2Dephasing`, with sliders for the position of one "weak" node and its T2, plus desktop GLMakie and web WGLMakie versions, a README, and smoke tests under `test/examples/` per the examples workflow. It would not touch library code.

Two things before I start: (1) is this distinct enough from existing examples and unclaimed? (2) I work with an agent-assisted workflow (Myncellium): the design and physics choices and the final code review are human-verified, but parts of the implementation were produced with AI assistance — if that is disqualifying for this program, please tell me now and I won't waste your time. Happy to move the discussion to a discussion post and do PR review during office hours.
